File: crates/relayer/src/event_generator.rs
use crate::config::RelayPair;
use crate::types::{ChainConfig, EventMeta, RelayEvent};
use anyhow::anyhow;
use anyhow::{Context, Result};
use ethers::{
    abi::{self},
    core::types::{Address, Bytes, H256, U256},
    prelude::*,
    providers::{Http, Provider},
    signers::{LocalWallet, Signer},
    utils::keccak256,
};
use std::collections::HashMap;
use std::{str::FromStr, sync::Arc, time::Duration};
use tokio::{sync::mpsc, time};
use tracing::{debug, error, info, instrument};

pub struct EventGenerator {
    chains: HashMap<u64, ChainConfig>,
    relay_pairs: Vec<RelayPair>,
    private_key: String,
    polling_interval: Duration,
    event_tx: mpsc::Sender<RelayEvent>,
}

impl EventGenerator {
    pub fn new(
        chains: HashMap<u64, ChainConfig>,
        relay_pairs: Vec<RelayPair>,
        private_key: String,
        polling_interval: Duration,
        event_tx: mpsc::Sender<RelayEvent>,
    ) -> Self {
        Self {
            chains,
            private_key,
            polling_interval,
            event_tx,
            relay_pairs,
        }
    }

    #[instrument(skip(self), name = "event_generator_start")]
    pub async fn start(&self) -> Result<()> {
        info!("Starting event generator");

        let mut interval_timer = time::interval(self.polling_interval);

        loop {
            interval_timer.tick().await;
            if let Err(e) = self.check_all_chains().await {
                error!(error = %e, "Error checking chains");
            }
        }
    }

    #[instrument(skip(self))]
    async fn check_all_chains(&self) -> Result<()> {
        for relay_pair in &self.relay_pairs {
            let source_chain = self
                .chains
                .get(&relay_pair.source_chain_id)
                .ok_or_else(|| {
                    anyhow::anyhow!(
                        "Source chain {} not found in config",
                        relay_pair.source_chain_id
                    )
                })?;

            let dest_chain = self.chains.get(&relay_pair.dest_chain_id).ok_or_else(|| {
                anyhow::anyhow!(
                    "Destination chain {} not found in config",
                    relay_pair.dest_chain_id
                )
            })?;

            match self
                .check_cross_chain_events(source_chain, dest_chain, relay_pair)
                .await
            {
                Ok(_) => {}
                Err(e) => error!(
                    source_chain = %source_chain.name,
                    dest_chain = %dest_chain.name,
                    error = %e,
                    "Error checking cross-chain events"
                ),
            }
        }
        Ok(())
    }

    #[instrument(skip(self), fields(source_chain = %source_chain.name, dest_chain = %dest_chain.name))]
    async fn check_cross_chain_events(
        &self,
        source_chain: &ChainConfig,
        dest_chain: &ChainConfig,
        relay_pair: &RelayPair,
    ) -> Result<()> {
        info!("Checking cross-chain events");

        // Connect to provider
        let provider = Provider::<Http>::try_from(&source_chain.rpc_url).context(format!(
            "Failed to create provider for {}",
            source_chain.name
        ))?;
        let client = Arc::new(provider);

        // Create wallet
        let wallet = LocalWallet::from_str(&self.private_key)
            .context("Failed to create wallet")?
            .with_chain_id(source_chain.chain_id);
        let client = SignerMiddleware::new(client, wallet);

        // Create resolver contract interface
        let resolver_address = Address::from_str(&relay_pair.source_resolver_address)
            .context("Invalid resolver address")?;

        // Create ABI for the cross-chain resolver interface
        let resolver_abi = abi::parse_abi(&[
            "function crossChainChecker(uint32 destinationChainId) external view returns (bool canExec, bytes memory execPayload, uint256 nonce)"
        ])?;
        let resolver_contract =
            Contract::new(resolver_address, resolver_abi, Arc::new(client.clone()));

        debug!("Calling crossChainChecker() on resolver");

        // Call the crossChainChecker function
        let dest_chain_id_u32 = dest_chain.chain_id as u32;
        let result: (bool, Bytes, U256) = resolver_contract
            .method("crossChainChecker", dest_chain_id_u32)?
            .call()
            .await?;

        let (can_exec, exec_payload, nonce) = result;

        if can_exec {
            info!(
                nonce = nonce.as_u64(),
                source_chain = source_chain.name,
                dest_chain = dest_chain.name,
                "✅ Cross-chain execution needed"
            );

            // Process the cross-chain event
            let tx_hash = self
                .request_remote_execution(&source_chain, relay_pair)
                .await?;

            // Extract event details and create the RelayEvent
            let event = self
                .extract_event_details(
                    tx_hash,
                    source_chain,
                    dest_chain,
                    exec_payload,
                    nonce.as_u64(),
                    relay_pair,
                )
                .await?;

            // Send the event to the proof fetcher
            if let Err(e) = self.event_tx.send(event).await {
                error!(error = %e, "Failed to send event to proof fetcher");
            }
        } else {
            debug!("⏳ No cross-chain execution needed");
        }

        Ok(())
    }

    #[instrument(skip(self), fields(source_chain = %source_chain.name, dest_chain = %destination_chain.name))]
    async fn extract_event_details(
        &self,
        tx_hash: H256,
        source_chain: &ChainConfig,
        destination_chain: &ChainConfig,
        exec_payload: Bytes,
        nonce: u64,
        relay_pair: &RelayPair,
    ) -> Result<RelayEvent> {
        // Get the transaction receipt to extract event details
        let provider = Provider::<Http>::try_from(&source_chain.rpc_url).context(format!(
            "Failed to create provider for {}",
            source_chain.name
        ))?;
        let tx_receipt = provider
            .get_transaction_receipt(tx_hash)
            .await?
            .ok_or_else(|| anyhow::anyhow!("Transaction receipt not found"))?;

        // Find the CrossChainExecRequested event in the logs
        let cross_chain_event = tx_receipt
            .logs
            .iter()
            .find(|log| {
                // Check if this log is from our source resolver address
                let from_resolver = log.address
                    == Address::from_str(&relay_pair.source_resolver_address).unwrap_or_default();

                // Check if the log has the CrossChainExecRequested event signature
                // Event: CrossChainExecRequested(uint32 indexed destinationChainId, bytes execPayload, uint256 indexed nonce)
                // Keccak256 hash of the event signature
                let event_signature = "CrossChainExecRequested(uint32,bytes,uint256)";
                let event_signature_hash = keccak256(event_signature.as_bytes());

                from_resolver
                    && log
                        .topics
                        .get(0)
                        .map_or(false, |t| t.as_bytes() == &event_signature_hash[..])
            })
            .ok_or_else(|| {
                anyhow::anyhow!("CrossChainExecRequested event not found in transaction")
            })?;

        // Create a relay event with actual transaction details
        let event = RelayEvent {
            source_chain: source_chain.clone(),
            source_resolver_address: relay_pair.source_resolver_address.clone(),
            destination_chain: destination_chain.clone(),
            dest_dapp_address: relay_pair.dest_dapp_address.clone(),
            exec_payload,
            nonce,
            meta: EventMeta {
                tx_hash: Some(tx_hash),
                block_number: tx_receipt
                    .block_number
                    .map(|n| n.as_u64())
                    .ok_or(anyhow!("block_number not found from receipt"))?,
                tx_index: tx_receipt.transaction_index.as_u32(),
                log_index: cross_chain_event
                    .log_index
                    .map(|n| n.as_u32())
                    .ok_or(anyhow!(
                        "log_index not found from CrossChainExecRequested event"
                    ))?,
            },
        };

        Ok(event)
    }

    async fn request_remote_execution(
        &self,
        source_chain: &ChainConfig,
        relay_pair: &RelayPair,
    ) -> Result<H256> {
        info!("Requesting remote execution");

        // Connect to provider
        let provider = Provider::<Http>::try_from(&source_chain.rpc_url).context(format!(
            "Failed to create provider for {}",
            source_chain.name
        ))?;
        let client = Arc::new(provider);

        // Create wallet
        let wallet = LocalWallet::from_str(&self.private_key)
            .context("Failed to create wallet")?
            .with_chain_id(source_chain.chain_id);
        let client = SignerMiddleware::new(client, wallet);

        // Create resolver contract interface
        let resolver_address = Address::from_str(&relay_pair.source_resolver_address)
            .context("Invalid resolver address")?;

        // Create ABI for the cross-chain resolver interface
        let resolver_abi = abi::parse_abi(&[
            "function requestRemoteExecution(uint32 destinationChainId) external",
        ])?;
        let resolver_contract =
            Contract::new(resolver_address, resolver_abi, Arc::new(client.clone()));

        // Call requestRemoteExecution
        info!("Calling requestRemoteExecution on resolver");
        let tx_req = resolver_contract
            .method::<_, ()>("requestRemoteExecution", relay_pair.dest_chain_id)?;
        let tx = tx_req.send().await?;

        let tx_hash = tx.tx_hash();
        info!(?tx_hash, "Transaction sent");

        // Wait for transaction to be mined
        let receipt = tx
            .await?
            .ok_or_else(|| anyhow::anyhow!("Transaction receipt not found"))?;

        info!(?receipt, "Transaction confirmed");

        Ok(tx_hash)
    }
}

File: ts-relayer/src/services/EventGenerator.ts
import { ethers } from 'ethers';
import { ChainConfig, RelayPair } from '../config/types';
import { logger } from '../utils/logger';
import { RelayEvent, EventMeta } from '../types';

// ABI fragments for the cross-chain resolver interface
const RESOLVER_ABI = [
  'function crossChainChecker(uint32 destinationChainId) external view returns (bool canExec, bytes memory execPayload, uint256 nonce)',
  'function requestRemoteExecution(uint32 destinationChainId) external',
  'event CrossChainExecRequested(uint32 indexed destinationChainId, bytes execPayload, uint256 indexed nonce)'
];

export class EventGenerator {
  private chains: Record<number, ChainConfig>;
  private relayPairs: RelayPair[];
  private privateKey: string;
  private pollingIntervalMs: number;
  private eventCallback: (event: RelayEvent) => Promise<void>;
  private running: boolean = false;

  constructor(
    chains: Record<number, ChainConfig>,
    relayPairs: RelayPair[],
    privateKey: string,
    pollingIntervalMs: number,
    eventCallback: (event: RelayEvent) => Promise<void>
  ) {
    this.chains = chains;
    this.relayPairs = relayPairs;
    this.privateKey = privateKey;
    this.pollingIntervalMs = pollingIntervalMs;
    this.eventCallback = eventCallback;
  }

  public async start(): Promise<void> {
    logger.info('Starting event generator');
    this.running = true;
    
    while (this.running) {
      try {
        await this.checkAllChains();
      } catch (error) {
        logger.error({ error }, 'Error checking chains');
      }
      
      // Wait for the polling interval
      await new Promise(resolve => setTimeout(resolve, this.pollingIntervalMs));
    }
  }

  public stop(): void {
    logger.info('Stopping event generator');
    this.running = false;
  }

  private async checkAllChains(): Promise<void> {
    for (const relayPair of this.relayPairs) {
      const sourceChain = this.chains[relayPair.sourceChainId];
      const destChain = this.chains[relayPair.destChainId];

      if (!sourceChain) {
        logger.error(`Source chain ${relayPair.sourceChainId} not found in config`);
        continue;
      }

      if (!destChain) {
        logger.error(`Destination chain ${relayPair.destChainId} not found in config`);
        continue;
      }

      try {
        await this.checkCrossChainEvents(sourceChain, destChain, relayPair);
      } catch (error) {
        logger.error({
          sourceChain: sourceChain.name,
          destChain: destChain.name,
          error
        }, 'Error checking cross-chain events');
      }
    }
  }

  private async checkCrossChainEvents(
    sourceChain: ChainConfig,
    destChain: ChainConfig,
    relayPair: RelayPair
  ): Promise<void> {
    logger.info({
      sourceChain: sourceChain.name,
      destChain: destChain.name
    }, 'Checking cross-chain events');

    // Connect to provider
    const provider = new ethers.providers.JsonRpcProvider(sourceChain.rpcUrl);
    
    // Create wallet
    const wallet = new ethers.Wallet(this.privateKey, provider);
    
    // Create resolver contract interface
    const resolverAddress = ethers.utils.getAddress(relayPair.sourceResolverAddress);
    const resolverContract = new ethers.Contract(resolverAddress, RESOLVER_ABI, wallet);

    logger.debug('Calling crossChainChecker() on resolver');

    // Call the crossChainChecker function
    const destChainIdU32 = destChain.chainId;
    const [canExec, execPayload, nonce] = await resolverContract.crossChainChecker(destChainIdU32);

    if (canExec) {
      logger.info({
        nonce: nonce.toString(),
        sourceChain: sourceChain.name,
        destChain: destChain.name
      }, '✅ Cross-chain execution needed');

      // Process the cross-chain event
      const txHash = await this.requestRemoteExecution(sourceChain, destChain, relayPair);

      // Extract event details and create the RelayEvent
      const event = await this.extractEventDetails(
        txHash,
        sourceChain,
        destChain,
        execPayload,
        nonce.toString(),
        relayPair
      );

      // Send the event to the proof fetcher via callback
      await this.eventCallback(event);
    } else {
      logger.debug('⏳ No cross-chain execution needed');
    }
  }

  private async requestRemoteExecution(
    sourceChain: ChainConfig,
    destChain: ChainConfig,
    relayPair: RelayPair
  ): Promise<string> {
    logger.info('Requesting remote execution');

    // Connect to provider
    const provider = new ethers.providers.JsonRpcProvider(sourceChain.rpcUrl);
    
    // Create wallet
    const wallet = new ethers.Wallet(this.privateKey, provider);
    
    // Create resolver contract interface
    const resolverAddress = ethers.utils.getAddress(relayPair.sourceResolverAddress);
    const resolverContract = new ethers.Contract(resolverAddress, RESOLVER_ABI, wallet);

    // Call requestRemoteExecution
    logger.info('Calling requestRemoteExecution on resolver');
    const tx = await resolverContract.requestRemoteExecution(destChain.chainId);

    logger.info({ txHash: tx.hash }, 'Transaction sent');

    // Wait for transaction to be mined
    const receipt = await tx.wait();
    logger.info({ 
      txHash: receipt.transactionHash,
      blockNumber: receipt.blockNumber
    }, 'Transaction confirmed');

    return receipt.transactionHash;
  }

  private async extractEventDetails(
    txHash: string,
    sourceChain: ChainConfig,
    destChain: ChainConfig,
    execPayload: string,
    nonce: string,
    relayPair: RelayPair
  ): Promise<RelayEvent> {
    // Get the transaction receipt to extract event details
    const provider = new ethers.providers.JsonRpcProvider(sourceChain.rpcUrl);
    const txReceipt = await provider.getTransactionReceipt(txHash);

    if (!txReceipt) {
      throw new Error('Transaction receipt not found');
    }

    // Create ABI interface for parsing logs
    const iface = new ethers.utils.Interface(RESOLVER_ABI);
    
    // Find the CrossChainExecRequested event in the logs
    const resolverAddress = ethers.utils.getAddress(relayPair.sourceResolverAddress);
    const eventSignature = 'CrossChainExecRequested(uint32,bytes,uint256)';
    const eventTopic = ethers.utils.id(eventSignature);
    
    const crossChainEvent = txReceipt.logs.find(log => 
      log.address.toLowerCase() === resolverAddress.toLowerCase() && 
      log.topics[0] === eventTopic
    );

    if (!crossChainEvent) {
      throw new Error('CrossChainExecRequested event not found in transaction');
    }

    // Create a relay event with actual transaction details
    const event: RelayEvent = {
      sourceChain: sourceChain,
      sourceResolverAddress: relayPair.sourceResolverAddress,
      destinationChain: destChain,
      destDappAddress: relayPair.destDappAddress,
      execPayload: execPayload,
      nonce: nonce,
      meta: {
        txHash: txHash,
        blockNumber: txReceipt.blockNumber,
        txIndex: txReceipt.transactionIndex,
        logIndex: crossChainEvent.logIndex,
      },
    };

    return event;
  }
}
